README.md
# realtor
Search realtor and do some filtering


### API Reference/Options

* `Culture` - `1` for EN, `2` for FR. Defaulted to 1.
* `ApplicationId` - Unused. Defaulted to 1.
* `PropertySearchTypeId`- Defaulted to 1. Determines the type of property, possible values:
    * `0` No Preference
    * `1` Residential
    * `2` Recreational
    * `3` Condo/Strata
    * `4` Agriculture
    * `5` Parking
    * `6` Vacant Land
    * `8` Multi Family

*Most useful options*

* `PriceMin` - Defaults to 0
* `PriceMax`
* `LongitudeMin` - Designates the bounds of the query, easiest to find these values from browser requests.
* `LongitudeMax`
* `LatitudeMin`
* `LatitudeMax`
* `TransactionTypeId`- Defaults to 2?
    * `1` For sale or rent
    * `2` For sale
    * `3` For rent
* `StoreyRange` - ``"min-max"`` i.e. `"2-3"`
* `BedRange` - `"min-max"`
* `BathRange` - `"min-max"`

*Others*

* Sorting:

Type | `SortBy` | `SortOrder`
---- | -------- | -----------
Low to High ($) | `1` | `A`
High to Low ($) | `1` | `D`
Date Posted: New to Old | `6` | `D`
Date Posted: Old to New | `6` | `A`
Open Houses First | `12` | `D`
More Photos First | `13` | `D`
Virtual Tour First | `11` | `D`

* `viewState` - `m`, `g`, or `1`. Seems irrelevant.
* `Longitude` - Longitude to focus on? Unneeded
* `Latitude` - Latitude to focus on? Unneeded
* `ZoomLevel` - not sure what this does
* `CurrentPage` - read somewhere that it maxes at 51
* `RecordsPerPage`
* `MaximumResults`
* `PropertyTypeGroupID` - ???
* `OwnershipTypeGroupId`
    * `0` Any
    * `1` Freehold
    * `2` Condo/Strata
    * `3` Timeshare/Fractional
    * `4` Leasehold
* `ViewTypeGroupId`
    * `0` Any
    * `1` City
    * `2` Mountain
    * `3` Ocean
    * `4` Lake
    * `5` River
    * `6` Ravine
    * `7` Other
    * `8` All Water Views
* `BuildingTypeId`
    * `0` Any
    * `1` House
    * `2` Duplex
    * `3` Triplex
    * `5` Residential Commercial Mix
    * `6` Mobile Home
    * `12` Special Purpose
    * `14` Other
    * `16` Row/Townhouse
    * `17` Apartment
    * `19` Fourplex
    * `20` Garden Home
    * `27` Manufactured Home/Mobile
    * `28` Commercial Apartment
    * `29` Manufactured Home
* `ConstructionStyleId`
    * `0` Any
    * `1` Attached
    * `3` Detached
    * `5` Semi-detached
    * `7` Stacked
    * `9` Link
* `AirCondition`- `0` or `1`, defaults 0
* `Pool` - `0` or `1`, defaults 0
* `Fireplace` - `0` or `1`, defaults 0
* `Garage` - `0` or `1`, defaults 0
* `Waterfront` - `0` or `1`, defaults 0
* `Acreage` - `0` or `1`, defaults 0
* `Keywords` - search text


CHANGELOG.md
Go train stations ,walk score,average income
Support filtering by distance between lot and subway stations 2017-08-24
Support filtering by orientation (use google map api) 2017-08-20
Support filtering by address name 2017-07-24


test.php
<?php
define('__ROOT__', dirname(__FILE__));
require_once(__ROOT__.'/lib/road.php');
require_once(__ROOT__.'/lib/angle.php');


$lat =(double) $argv[1];
$lng =(double) $argv[2];

$angle = 0.0;
//$point = get_snapped_point($lat, $lng);
//$angle = getRotateAngle($lat, $lng, $point["latitude"], $point["longitude"]);
//Var_dump($angle);

/*
$api_key="123";
$url_arr = array("https://roads.googleapis.com/v1/nearestRoads?points=$lat,$lng&key=".$api_key,
                 "https://roads.googleapis.com/v1/snapToRoads?path=$lat,$lng&key=".$api_key);

$url= $url_arr[array_rand($url_arr)];
var_dump($url) ;

$add_txt_array= explode ("|", "15 WERTHEIM COURT UNIT 302|RICHMOND HILL, ON L4B3H7");
$txt_array= explode( " ",$add_txt_array[0]);
$dst_array= array_slice ($txt_array, 2);
var_dump(implode (" " ,$dst_array));
*/

/*
$url = "https://en.wikipedia.org/wiki/List_of_Toronto_subway_stations";
$page = file_get_contents($url, false);
//var_dump (trim ($page));

preg_match_all('/<th scope="row"><a href="(.*?)" title=(.*?)>(.*?)<\/a><\/th>/', trim($page), $matches);

$final=array();
foreach ($matches[1] as $key=> $sub_url)
{
    $str = file_get_contents ("https://en.wikipedia.org".$sub_url, false);
    preg_match ('/<span class="geo">(.*?)<\/span>/', trim($str), $geo);
    array_push ($final ,array ("name"=>$matches[3][$key], "geo" => $geo[1]));
    
}

file_put_contents("lib/stations.json",json_encode($final));
*/


function address ($address,...$others)
{
    
    echo( $address);
    return false;
}

/*
filter by orientation
for north-faceing angle should be between 90 to 270
*/
function orientation ($address,$lat, $lng)
{
    echo("o");
    return false;


}

//filter by nearest subway stations
function subway ($address,$lat, $lng)
{
    echo("s");
    return false;
    
}


function filter ($address,$lat, $lng)
{
    $filter_arr= array ("address"=>false, "subway"=> false, "orientation"=> true);
    $ret= true;
    foreach($filter_arr as $key=> $value)
    {
        if ($value)
        {
            $ret &= call_user_func($key, $address,$lat,$lng);    
        }        
    }
    return $ret;
}

filter ("awwwwwtony",1 ,2);

?>



lib/road.php
<?php
/*
 Give a lot's Geo info (lat lng)
 Use google map road api to get snapped points (lat1 lng1) on the nearlest road of the given lot
 Use vector (lng1 lat1)--->(lng1 lat1) to caculate orientation
*/

function get_snapped_point ($lat, $lng)
{
    $options = array(
        'http' => array(
            'header'  => "Content-type: application/json",
            'method'  => 'GET' 
        )
    );
    
    //ATTENTION :Replace your own api_key
    $api_key="key";
    $url_arr = array("https://roads.googleapis.com/v1/nearestRoads?points=$lat,$lng&key=".$api_key,
                     "https://roads.googleapis.com/v1/snapToRoads?path=$lat,$lng&key=".$api_key);

    $url= $url_arr[array_rand($url_arr)];
    $context  = stream_context_create($options);

    $result = file_get_contents($url, false, $context);
    $result_arr=json_decode($result,true);

    
    return ($result_arr["snappedPoints"][0]["location"]);
}

//var_dump(get_snapped_point (43.8006949,-79.4455814));

?>


lib/angle.php
<?php
/*
  For vectors p1 and p2
  Dot product
  (p1^p2)/(|p1|*|p2|) = cos(theta)，0<=theta<=180
  
  Cross product
  p1 * p2 = x1y2  - x2 y1 = -p2 * p1
  if p1 * p2 is positive, then p1 is clockwise from p2 with respect to the origin (0, 0);
  if this cross product is negative, then p1 is counterclockwise from p2.

  Use (lng lat) as poit (x y)
*/


function getRotateAngle($lat1, $lng1, $lat2, $lng2)
{
   $epsilon = 1.0e-6;
    $nyPI = acos(-1.0);

    //vector p1 ($lat2, $lng2)--->($lat1, $lng1) 
    $x1 = $lng1-$lng2;
    $y1 = $lat1-$lat2;

    //vector p2 (0.0,0.0)---->(0.0,1.0) stands North
    $x2 =0.0;
    $y2 =1.0;
 
    // normalize
    $dist1 = sqrt( $x1 * $x1 + $y1 * $y1 );
    if (0.0 === $dist1)
    {
        echo ("Overlap points $lat1, $lng1, $lat2, $lng2");
        return 0.0;
    }
    $x1 /= $dist1;
    $y1 /= $dist1;
    $dist2 = sqrt( $x2 * $x2 + $y2 * $y2 );
    $x2 /= $dist2;
    $y2 /= $dist2;
    // dot product
    $dot = $x1 * $x2 + $y1 * $y2;
    if (abs($dot-1.0) <= $epsilon )
        $angle = 0.0;
    else if ( abs($dot+1.0) <= $epsilon )
        $angle = $nyPI;
    else {
        
        $angle = acos($dot);
        //cross product
        $cross = $x1 * $y2 - $x2 * $y1;
        // vector p2 is clockwise from vector p1
        // with respect to the origin (0.0)
        if ($cross < 0 ) {
            $angle = 2 * $nyPI - $angle;
        }    
    }
    $degree = $angle *  180.0 / $nyPI;
    return $degree;
}


//var_dump(getRotateAngle(43.8006949,-79.4455814 ,43.800668646930951, -79.445804597379308));
?>


lib/distance.php
<?php
/*
 Calculate transit distance (seconds) between your lot and subway stations
 Have limits to call google map api
*/
define('__DIR__', dirname(__FILE__));

function googleDistance ($lat1, $lng1, $lat2, $lng2, $mode)
{
    $options = array(
        'http' => array(
            'header'  => "Content-type: application/json",
            'method'  => 'GET' 
        )
    );
    
    //ATTENTION :Replace your own api_key
    $api_key="key";
    $url= "https://maps.googleapis.com/maps/api/directions/json?mode=$mode&origin=$lat1,$lng1&destination=$lat2,$lng2&key=".$api_key;
    $context  = stream_context_create($options);

    $result = file_get_contents($url, false, $context);
    $result_arr=json_decode($result,true);

    
    return ($result_arr["routes"][0]["legs"][0]["duration"]["value"]);
}

/*

Calculate distance between two geo points
*/

function distance ($lat1, $lng1, $lat2 ,$lng2)
{
    $p = 0.017453292519943295;
    $a = 0.5 - cos(($lat2 - $lat1) * $p)/2 + cos($lat1 * $p) * cos($lat2 * $p) * (1 - cos(($lng2 - $lng1) * $p)) / 2;
    return 12742 * asin(sqrt($a));
}

/*
  Get nearest subway station
*/
function getStation ($lat ,$lng)
{
    $stations_arr =json_decode(file_get_contents (__DIR__.'/stations.json') ,true);
  
    $distance=1e6;
    $k=0;
    foreach ($stations_arr as $key => $station)
    {
        $loc= explode(";",$station["geo"]);
        $dist = distance ($lat ,$lng, (double) trim($loc[0]), (double) trim($loc[1]));
        if ($distance > $dist)
        {
            $distance = $dist;
            $k= $key;   
        }
    }
    $ret = $stations_arr[$k];
    $ret ["distance"] =$distance;
    return $ret;
}

//var_dump(distance (43.8006949,-79.4455814,43.78056,-79.41472));

//var_dump(getstation (43.800927,-79.4343206));
?>

realtor.php
<?php
/*
This is a tool filtering houses in Toronto
Run like this: php -f realor.php
*/
define('__ROOT__', dirname(__FILE__));
require_once(__ROOT__.'/lib/road.php');
require_once(__ROOT__.'/lib/angle.php');
require_once(__ROOT__.'/lib/distance.php');

$url = "https://api2.realtor.ca/Listing.svc/PropertySearch_Post";
$params = array ("CultureId" => "1",
                 "ApplicationId" => "1", 
                 "ParkingSpaceTotal" => "4",
                 "RecordsPerPage" => "9",
                 "BuildingTypeId" => "1",
                 "ConstructionStyleId" => "3",
                 "PriceMin" => "1200000",
                 "PriceMax" => "1750000",
                 "MaximumResults" => "9",
                 "PropertySearchTypeId" => "1",
                 "TransactionTypeId" => "2",
                 "StoreyRange" => "0-0",
                 "BedRange" => "3-0",
                 "BathRange" => "2-0",
                 "LongitudeMin" => "-79.65129996230468",
                 "LongitudeMax" => "-79.23862601210936",
                 "LatitudeMin" => "43.702077668041646",
                 "LatitudeMax" =>"43.923312715190676",
                 "SortOrder" => "D",
                 "SortBy" => "6",
                 "viewState" => "m",
                 "favouritelistingids" => "18347321",
                 "CurrentPage" => "1",
                 "ZoomLevel" => "12",
                 "PropertyTypeGroupID" => "1",
                 "Version" => "6.0" );
$data = http_build_query ($params);
$data_len = strlen ($data);
$opts = array('http' =>
              array(
                  'method'  => 'POST',
                  'header'  => "Content-Type: application/x-www-form-urlencoded\r\nContent-Length: $data_len\r\n",
                  'content' => $data 
              )
);

$context  = stream_context_create($opts);
$result = file_get_contents($url, false, $context);

$house_arr=array ();

/*
filter by street type name
For example CRT,PL,CRES
Generally,houses locate on these streets will be quiet and beautiful
*/
function address ($address,...$others)
{
    //get rid of sreet No.,street name,focus on street type
    $add_arr= explode ("|", $address);
    $txt_array= explode ( " ",$add_arr[0]);
    $dst_array= array_slice($txt_array, 2);
    $address_str= implode (" " ,$dst_array);
    
    $address_arr= array ("PL","CRT","CRES","LANE","CIRC");
    $ret=false;
    
    foreach ($address_arr as $add){
        if (strpos($address_str, $add))
        {
            $ret=true;
            break;
        }
    }
    return $ret;
}

/*
filter by orientation
for north-faceing angle should be between 90 to 270
*/
function orientation ($address,$lat, $lng)
{
    $angle=0.0;
    $point = get_snapped_point($lat, $lng);
    $angle = getRotateAngle($lat, $lng, $point["latitude"], $point["longitude"]);
    return (($angle >105) && ($angle<255));
}

//filter by nearest subway stations
function subway ($address,$lat, $lng)
{
    $ret = false;
    $station = getStation ($lat ,$lng);
    if ($station["distance"] < 1.0)
    {
        $ret = true;
    }
    return $ret;
}


function filter ($address,$lat, $lng)
{
    $filter_arr= array ("address"=>false, "subway"=> true, "orientation"=> true);
    $ret= true;
    foreach($filter_arr as $key=> $value)
    {
        if ($value)
        {
            $ret &= call_user_func($key, $address,$lat,$lng);
        }
    }
    return $ret;
}

$result_arr=json_decode($result,true);
//var_dump ($result_arr);
//exit();

if(!empty($result_arr)) {

    foreach ($result_arr["Results"] as $house){
        $lat = (double)$house["Property"]["Address"]["Latitude"];
        $lng = (double)$house["Property"]["Address"]["Longitude"];
        $add = $house["Property"]["Address"]["AddressText"];

        if (filter ($add,$lat,$lng))
        {
            array_push($house_arr,$house["MlsNumber"]);
        }            
    }
    
    for ($page=2; $page<=$result_arr["Paging"]["TotalPages"]; $page++)
    {
        $params["CurrentPage"]=$page;
        $data = http_build_query ($params);
        $data_len = strlen ($data);
        $opts = array('http' =>
                      array(
                          'method'  => 'POST',
                          'header'  => "Content-Type: application/x-www-form-urlencoded\r\nContent-Length: $data_len\r\n",
                          'content' => $data 
                      )
        );

        $context  = stream_context_create($opts);
        $ret = file_get_contents($url, false, $context);
        $ret_arr=json_decode($ret,true);
        foreach ($ret_arr["Results"] as $house){
            $lat = (double)$house["Property"]["Address"]["Latitude"];
            $lng = (double)$house["Property"]["Address"]["Longitude"];
            $add = $house["Property"]["Address"]["AddressText"];

            if (filter ($add,$lat,$lng))
            {
                array_push($house_arr,$house["MlsNumber"]);
            }            
        }        
    }
}


var_dump($house_arr);

?>
